software/contrib/beat_writer.py

```diff
@@ -0,0 +1,165 @@
+class BeatWriter(EuroPiScript):
+    def __init__(self):
+        super().__init__()
+        
+        self.length = 16
+        
+        state = self.load_state_json()
+        self.rhythm_1 = state.get("rhythm_1", self.create_blank_rhythm(self.length))
+        self.rhythm_2 = state.get("rhythm_2", self.create_blank_rhythm(self.length))
+        
+        self.previous_step_ticks = ticks_ms
+        self.previous_step_length = 0
+    
+        self.step = 0
+                
+        self.din_out = cv1
+        self.rhythm_1_output = cv2
+        self.rhythm_2_output = cv5
+        self.start_output = cv4
+        
+        self.clock_input = din
+        self.restart_input = DigitalInput(26)
+        
+        self.clock_input.handler(self.increment_step)
+        self.restart_input.handler(self.restart_rhythms)
+        
+        
+        b1.handler(self.place_beat)
+        b2.handler(self.place_beat)
+
+        b1.handler_falling(self.check_if_reset_rhythm_1)
+        b2.handler_falling(self.check_if_reset_rhythm_2)
+
+    @classmethod
+    def display_name(cls):
+        return "Beat Writer"
+
+    def create_blank_rhythm(self, length):
+        rhythm = []
+        for beat in range(length):
+            rhythm.append(0)
+        return rhythm
+    
+    def update_randomisation(self):
+        self.randomisation_1 = k1.read_position()
+        self.randomisation_2 = k2.read_position()
+    
+    def update_outputs(self):
+        self.update_randomisation()
+        output_1_value = randint(0, 1) if randint(2, 98) < self.randomisation_1 else self.rhythm_1[self.step]
+        output_2_value = randint(0, 1) if randint(2, 98) < self.randomisation_2 else self.rhythm_2[self.step]
+        
+        self.rhythm_1_output.value(output_1_value)
+        self.rhythm_2_output.value(output_2_value)
+    
+    def increment_step(self):
+        current_ticks = ticks_ms()
+        self.previous_step_length = ticks_diff(current_ticks, self.previous_step_ticks)
+        self.previous_step_ticks = current_ticks
+        
+        self.start_output.value(0)
+        self.step += 1
+        if self.step == self.length:
+            self.step = 0
+            self.start_output.value(1)
+            
+        self.update_outputs()
+        self.update_display()
+        
+    def restart_rhythms(self):
+        self.step = 0
+        self.update_outputs()
+        self.update_display()
+        
+    def check_if_reset_rhythm_1(self):
+        current_ticks = ticks_ms()
+        b1_difference = ticks_diff(current_ticks, b1.last_pressed())
+        if b1_difference > 300:
+            self.reset_rhythm(self.rhythm_1)
+        self.save_state()
+            
+    def check_if_reset_rhythm_2(self):
+        current_ticks = ticks_ms()
+        b2_difference = ticks_diff(current_ticks, b2.last_pressed())
+        if b2_difference > 300:
+            self.reset_rhythm(self.rhythm_2)
+        self.save_state()
+            
+    def reset_rhythm(self, rhythm):
```

Comment:
[required] This function effectively duplicates the functionality of `create_blank_rhythm`.  I think that we only need one of them. If you got rid of this function, then the code in `check_if_reset_rhythm_2` could use:

```python
if b2_difference > 300:
    self.rhythm_2 = create_blank_rhythm(self.length)
```